### PR TITLE
Fix right hand side dict assignment for DataFrames

### LINKED ISSUE #9874: Dictionary assignment with loc with existing index doesn't properly match column names in dict
uses the example DataFrame from #9857.

```python
In [3]: df = pd.DataFrame({'x': [1,2,6], 'y': [2,2,8], 'z':[-5,0,5]})

In [4]: df = df.set_index('z')

In [5]: rhs = {'x': 9, 'y': 99}

In [6]: df.loc[5]
Out[6]:
x    6
y    8
Name: 5, dtype: int64

In [7]: df.loc[5] = rhs

In [8]: df
Out[8]:
                      x                    y
z
-5                    1                    2
 0                    2                    2
 5  {u'y': 99, u'x': 9}  {u'y': 99, u'x': 9}
```

### PR DESCRIPTION
**Pull request by cpcloud**
closes #9874

### CODE REVIEW

File: doc/source/indexing.rst [hunk 1]
```diff
@@ -249,6 +249,14 @@ new column.
 If you are using the IPython environment, you may also use tab-completion to
 see these accessible attributes.
 
+You can also assign a ``dict`` to a row of a ``DataFrame``:
```
Comment (jorisvandenbossche) [inline doc/source/indexing.rst]: Can you maybe specify here that the dict keys are matched with the column names?

**Comment by jreback** [comment]
@tjragan that's very odd to do anyhow. Supported for nested data structues is very limited, especially with indexing operations.

**Comment by tjragan** [comment]
It may be odd, but I have a fairly complex data analysis project that uses this extensively.  In any case, preserving pandas’  ability to store arbitrary python objects seems like a good idea, and this essentially removes the ability to have a key-value store in a cell.
---------------------------------------------------------
Dr. T.J. Ragan
Research Associate

Department of Biochemistry, University of Leicester
Henry Wellcome Building, Lancaster Road
Leicester LE1 9HN, United Kingdom

On 27 May 2015, at 11:57, jreback <notifications@github.com<mailto:notifications@github.com>> wrote:


@tjragan<https://github.com/tjragan> that's very odd to do anyhow. Supported for nested data structues is very limited, especially with indexing operations.

**Comment by jreback** [comment]
@tjragan what version of pandas are you using where this actually works? it doesn't work in >=0.14.1

**Comment by tjragan** [comment]
it broke when I installed 0.16.1, but works in 0.16.0

On 27 May 2015, at 20:56, jreback <notifications@github.com<mailto:notifications@github.com>> wrote:


@tjragan<https://github.com/tjragan> what version of pandas are you using where this actually works? it doesn't work in >=0.14.1

**Comment by jreback** [comment]
ok, see #10219 

pull-requests to fix are welcome

**Pull request merged**